Invalid version detection when tests present

Hey!

There is an invalid version detected with mingo on repo with tests . If i run `go test` by 1.16 version its work OK:
```
$ docker run -w /go/src -v .:/go/src golang:1.16 go test ./...
ok      github.com/alexdyukov/httpencoder   0.004s
```
but mingo tell me to upgrade go.mod up to 1.18 because testing.MainStart found:
```
$ go run github.com/bobg/mingo/cmd/mingo@latest -tests -check -v
/mnt/src/httpencoder/httpencoder.go:14:14: 7 ("context".Context)
/home/user/.cache/go-build/da/da4e65ca611572657c91072fb1a91ebac4181a9803104dff938df30632d6743c-d:57:7: 18 ("testing".MainStart)
Error scanning directory: go.mod declares version 1.16 but computed minimum is 1.18 [/home/user/.cache/go-build/da/da4e65ca611572657c91072fb1a91ebac4181a9803104dff938df30632d6743c-d:57:7: 18 ("testing".MainStart)]
exit status 1
```

Sure. Same error:
```
$ go run github.com/bobg/mingo/cmd/mingo@3ed5fa8 -tests -check -v
go: downloading github.com/bobg/mingo v0.9.7-0.20240705165823-3ed5fa8c1205
warning: GOPATH set to GOROOT (/home/user/go) has no effect
/mnt/c/Users/Lenovo/Downloads/FTA/src/httpencoder/decode.go:9:25: 3 ("sync".Pool)
/mnt/c/Users/Lenovo/Downloads/FTA/src/httpencoder/httpencoder.go:15:14: 7 ("context".Context)
/home/user/.cache/go-build/0e/0e70fa0823bc74d50197606763f245c57182455839ab3766d19b7bd229a94a14-d:57:7: 18 ("testing".MainStart)
Error scanning directory: go.mod declares version 1.16 but computed minimum is 1.18 [/home/user/.cache/go-build/0e/0e70fa0823bc74d50197606763f245c57182455839ab3766d19b7bd229a94a14-d:57:7: 18 ("testing".MainStart)]
exit status 1
```

I run `go install` and got same error:
```
$ go install github.com/bobg/mingo/cmd/mingo@latest
warning: GOPATH set to GOROOT (/home/user/go) has no effect
$ mingo -tests -check -v
/mnt/c/Users/Lenovo/Downloads/FTA/src/httpencoder/decode.go:9:25: 3 ("sync".Pool)
/mnt/c/Users/Lenovo/Downloads/FTA/src/httpencoder/httpencoder.go:15:14: 7 ("context".Context)
/home/user/.cache/go-build/0e/0e70fa0823bc74d50197606763f245c57182455839ab3766d19b7bd229a94a14-d:57:7: 18 ("testing".MainStart)
Error scanning directory: go.mod declares version 1.16 but computed minimum is 1.18 [/home/user/.cache/go-build/0e/0e70fa0823bc74d50197606763f245c57182455839ab3766d19b7bd229a94a14-d:57:7: 18 ("testing".MainStart)]
```

system library uses:
```
sync.Pool			go1.3
bytes.Buffer{} 			go1.0
context.Context			go1.7
io.Writer			go1.0
http.Request.Header.Get		go1.0
http.Handler.ServeHTTP		go1.0
bytes.Buffer.Bytes()		go1.0
http.DetectContentType		go1.0
http.ResponseWriter.WriteHeader	go1.0
http.Error			go1.0
io.NopCloser			go1.16
```
by cmd output i see mingo detects context.Context and sync.Pool, but missing calling io.NopCloser. I make a test code with same error. Feel free debugging with it:
```
user@DESKTOP-D8T7N02:~/src/benchmarks$ go run github.com/bobg/mingo/cmd/mingo@44c0f53 -check -v
warning: GOPATH set to GOROOT (/home/user/go) has no effect
/mnt/c/Users/Lenovo/Downloads/FTA/src/benchmarks/main.go:10:64: 1 (function body with no final return statement)
Error scanning directory: go.mod declares version 1.18 but computed minimum is 1.1 [/mnt/c/Users/Lenovo/Downloads/FTA/src/benchmarks/main.go:10:64: 1 (function body with no final return statement)]
exit status 1
user@DESKTOP-D8T7N02:~/src/benchmarks$ cat main.go
package main

import (
        "bytes"
        "fmt"
        "io"
)

func main() {
        fmt.Println(len(customReadAll(io.NopCloser(&bytes.Buffer{}))))
}

func customReadAll(reader io.Reader) []byte {
        buf := make([]byte, 100)
        retVal := []byte{}
        for {
                n, err := reader.Read(buf)

                retVal = append(retVal, buf...)
                buf = buf[:0]

                if n < 100 || err != nil {
                        return retVal
                }
        }

        return nil
}
```

--- a/_testdata/bug14/go.mod
+++ b/_testdata/bug14/go.mod
@@ -1,0 +1,3 @@
+module bug14
+
+go 1.18

--- a/_testdata/bug14/main.go
+++ b/_testdata/bug14/main.go
@@ -1,0 +1,32 @@
+// I am indebted to https://github.com/alexdyukov for this code
+// and the attendant bug report.
+// See https://github.com/bobg/mingo/issues/14
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+)
+
+func main() {
+	fmt.Println(len(customReadAll(io.NopCloser(&bytes.Buffer{}))))
+}
+
+func customReadAll(reader io.Reader) []byte {
+	buf := make([]byte, 100)
+	retVal := []byte{}
+	for {
+		n, err := reader.Read(buf)
+
+		retVal = append(retVal, buf...)
+		buf = buf[:0]
+
+		if n < 100 || err != nil {
+			return retVal
+		}
+	}
+
+	return nil
+}

--- a/expr.go
+++ b/expr.go
@@ -386,6 +386,10 @@ func (p *pkgScanner) callExpr(expr *ast.CallExpr) error {
 		return nil
 	}
 
+	return p.callArgs(expr)
+}
+
+func (p *pkgScanner) callArgs(expr *ast.CallExpr) error {
 	for _, arg := range expr.Args {
 		if err := p.expr(arg); err != nil {
 			return err
@@ -394,7 +398,6 @@ func (p *pkgScanner) callExpr(expr *ast.CallExpr) error {
 			return nil
 		}
 	}
-
 	return nil
 }
 
@@ -465,7 +468,8 @@ func (p *pkgScanner) builtinCall(expr *ast.CallExpr) error {
 		}
 		p.greater(result)
 	}
-	return nil
+
+	return p.callArgs(expr)
 }
 
 func getID(expr ast.Expr) *ast.Ident {

--- a/history_test.go
+++ b/history_test.go
@@ -31,6 +31,11 @@ func TestHistory(t *testing.T) {
 		pkgpath: "crypto/elliptic",
 		ident:   "Curve",
 		want:    0,
+	}, {
+		// https://github.com/bobg/mingo/issues/14
+		pkgpath: "io",
+		ident:   "NopCloser",
+		want:    16,
 	}}
 
 	for _, tc := range cases {

--- a/regression_test.go
+++ b/regression_test.go
@@ -1,0 +1,15 @@
+package mingo
+
+import "testing"
+
+// https://github.com/bobg/mingo/issues/14
+func TestBug14(t *testing.T) {
+	var s Scanner
+	res, err := s.ScanDir("_testdata/bug14")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if v := res.Version(); v != 16 {
+		t.Errorf("got %d, want 16", v)
+	}
+}

--- a/scan.go
+++ b/scan.go
@@ -226,15 +226,21 @@ func (s *Scanner) isMax() bool {
 	return s.Result.Version() >= s.h.max
 }
 
-var goCache = os.Getenv("GOCACHE")
-
 func isCacheFile(filename string) (bool, error) {
-	if goCache == "" {
-		return false, nil
+	cacheDir := os.Getenv("GOCACHE")
+	if cacheDir == "" {
+		var err error
+		cacheDir, err = os.UserCacheDir()
+		if err != nil {
+			return false, errors.Wrap(err, "getting user cache directory")
+		}
+		if cacheDir == "" {
+			return false, nil
+		}
 	}
-	rel, err := filepath.Rel(goCache, filename)
+	rel, err := filepath.Rel(cacheDir, filename)
 	if err != nil {
-		return false, errors.Wrapf(err, "computing relative path from %s to %s", goCache, filename)
+		return false, errors.Wrapf(err, "computing relative path from %s to %s", cacheDir, filename)
 	}
 	return !strings.HasPrefix(rel, "../"), nil
 }

--- a/scan.go
+++ b/scan.go
@@ -6,6 +6,7 @@ import (
 	"go/token"
 	"go/types"
 	"os"
+	"path/filepath"
 	"regexp"
 	"runtime"
 	"strconv"
@@ -142,6 +143,13 @@ func (s *Scanner) scanPackageHelper(pkgpath string, fset *token.FileSet, info *t
 
 	for _, file := range files {
 		filename := p.fset.Position(file.Pos()).Filename
+		isInCache, err := isCacheFile(filename)
+		if err != nil {
+			return errors.Wrapf(err, "checking whether %s is in GOCACHE", filename)
+		}
+		if isInCache {
+			continue
+		}
 		if err := p.file(file); err != nil {
 			return errors.Wrapf(err, "scanning file %s", filename)
 		}
@@ -216,4 +224,17 @@ func (s *Scanner) ensureHistory() error {
 // Prereq: e.ensureHistory has been called.
 func (s *Scanner) isMax() bool {
 	return s.Result.Version() >= s.h.max
+}
+
+var goCache = os.Getenv("GOCACHE")
+
+func isCacheFile(filename string) (bool, error) {
+	if goCache == "" {
+		return false, nil
+	}
+	rel, err := filepath.Rel(goCache, filename)
+	if err != nil {
+		return false, errors.Wrapf(err, "computing relative path from %s to %s", goCache, filename)
+	}
+	return !strings.HasPrefix(rel, "../"), nil
 }

